src/main/java/org/candlepin/subscriptions/marketplace/MarketplaceSubscriptionIdProvider.java

```diff
@@ -0,0 +1,107 @@
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.marketplace;
+
+import org.candlepin.subscriptions.db.SubscriptionRepository;
+import org.candlepin.subscriptions.db.model.ServiceLevel;
+import org.candlepin.subscriptions.db.model.Subscription;
+import org.candlepin.subscriptions.db.model.Usage;
+import org.candlepin.subscriptions.files.ProductProfile;
+import org.candlepin.subscriptions.files.ProductProfileRegistry;
+import org.candlepin.subscriptions.subscription.SubscriptionSyncController;
+import org.candlepin.subscriptions.tally.UsageCalculation.Key;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.util.Assert;
+
+import java.time.OffsetDateTime;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Class responsible for searching Swatch database for subscriptionIds corresponding to usage keys and if
+ * none is found, delegates fetching the subscriptionId to the {@link MarketplaceSubscriptionCollector}.
+ */
+@Component
+public class MarketplaceSubscriptionIdProvider {
+    private static final Logger log = LoggerFactory.getLogger(MarketplaceSubscriptionIdProvider.class);
+
+    private final MarketplaceSubscriptionCollector collector;
+    private final SubscriptionRepository subscriptionRepo;
+    private final SubscriptionSyncController syncController;
+    private final ProductProfileRegistry profileRegistry;
+
+    @Autowired
+    public MarketplaceSubscriptionIdProvider(MarketplaceSubscriptionCollector collector,
+        SubscriptionRepository subscriptionRepo, SubscriptionSyncController syncController,
+        ProductProfileRegistry profileRegistry) {
+        this.collector = collector;
+        this.subscriptionRepo = subscriptionRepo;
+        this.syncController = syncController;
+        this.profileRegistry = profileRegistry;
+    }
+
+    public Optional<Object> findSubscriptionId(String accountNumber, Key usageKey,
+        OffsetDateTime rangeStart, OffsetDateTime rangeEnd) {
+        Assert.isTrue(Usage._ANY != usageKey.getUsage(), "Usage cannot be _ANY");
+        Assert.isTrue(ServiceLevel._ANY != usageKey.getSla(), "Service Level cannot be _ANY");
+
+        String productId = usageKey.getProductId();
+        ProductProfile profile = profileRegistry.findProfileForSwatchProductId(productId);
+        Set<String> roles = profile.getRolesBySwatchProduct().getOrDefault(productId, Collections.emptySet());
+
+        List<Subscription> result =
+            subscriptionRepo.findSubscriptionByAccountAndUsageKey(accountNumber, usageKey, roles);
+        result = filterByDateRange(result, rangeStart, rangeEnd);
```

Comment:
@awood, I was just thinking... if we do one more filter here: filter out any subs without a marketplace subscription ID, it'll probably be a bit more robust. I'm thinking we might encounter a situation where there are some very similar looking non-marketplace subs we'd want to filter out.

Reply:
Updated.